evidence/reactor.go

```diff
@@ -21,233 +46,346 @@ const (
-		evR.Switch.StopPeerForError(src, err)
-		return
+// handleEvidenceMessage handles enevelopes sent from peers on the EvidenceChannel.
+// It returns an error only if the Envelope.Message is unknown for this channel
+// or if the given evidence is invalid. This should never be called outside of
+// handleMessage.
+func (r *Reactor) handleEvidenceMessage(envelope p2p.Envelope) error {
+	switch msg := envelope.Message.(type) {
+	case *tmproto.EvidenceList:
+		r.Logger.Debug("received evidence list", "num_evidence", len(msg.Evidence), "peer", envelope.From.String())
+
+		// TODO: Refactor the Evidence type to not contain a list since we only ever
+		// send and receive one piece of evidence at a time. Or potentially consider
+		// batching evidence.
+		for i := 0; i < len(msg.Evidence); i++ {
+			ev, err := types.EvidenceFromProto(&msg.Evidence[i])
+			if err != nil {
+				r.Logger.Error("failed to convert evidence", "peer", envelope.From.String(), "err", err)
+				continue
+			}
+
+			if err := r.evpool.AddEvidence(ev); err != nil {
+				r.Logger.Error("failed to add evidence", "peer", envelope.From.String(), "err", err)
+
+				// If we're given invalid evidence by the peer, notify the router that
+				// we should remove this peer by returning an error.
+				if _, ok := err.(*types.ErrInvalidEvidence); ok {
+					return err
+				}
+			}
+		}
+
+	default:
+		r.Logger.Error("received unknown message", "msg", msg, "peer", envelope.From.String())
+		return fmt.Errorf("received unknown message: %T", msg)
 	}
 
-	for _, ev := range evis {
-		err := evR.evpool.AddEvidence(ev)
-		switch err.(type) {
-		case *types.ErrInvalidEvidence:
-			evR.Logger.Error(err.Error())
-			// punish peer
-			evR.Switch.StopPeerForError(src, err)
+	return nil
+}
+
+// handleMessage handles an Envelope sent from a peer on a specific p2p Channel.
+// It will handle errors and any possible panics gracefully. A caller can handle
+// any error returned by sending a PeerError on the respective channel.
+func (r *Reactor) handleMessage(chID p2p.ChannelID, envelope p2p.Envelope) (err error) {
+	defer func() {
+		if e := recover(); e != nil {
+			err = fmt.Errorf("panic in processing message: %v", e)
+		}
+	}()
+
+	switch chID {
+	case EvidenceChannel:
+		err = r.handleEvidenceMessage(envelope)
+
+	default:
+		err = fmt.Errorf("unknown channel ID (%d) for envelope (%v)", chID, envelope)
+	}
+
+	return err
+}
+
+// processEvidenceCh implements a blocking event loop where we listen for p2p
+// Envelope messages from the evidenceCh.
+func (r *Reactor) processEvidenceCh() {
+	defer r.evidenceCh.Close()
+
+	for {
+		select {
+		case envelope := <-r.evidenceCh.In():
+			if err := r.handleMessage(r.evidenceCh.ID(), envelope); err != nil {
+				r.Logger.Error("failed to process envelope", "ch_id", r.evidenceCh.ID(), "envelope", envelope, "err", err)
+				r.evidenceCh.Error() <- p2p.PeerError{
+					PeerID:   envelope.From,
+					Err:      err,
+					Severity: p2p.PeerErrorSeverityLow,
```

Comment:
What does a low severity peer error do/mean?

Reply:
We're not really sure about the semantics here yet, we'll figure it out when we build the router. The idea is that a low severity might e.g. cause a disconnect, while a high severity might ban the peer entirely. But we might replace this with other concepts, or we might use different heuristics for this.

Reply:
Yeah I guess this is something else we can work out at a later point. It might not even be a static rule. Say for example if you're running low on active peers then you might still remain detected but in another circumstance where you are connected to too many then the slightest error could cause the router to disconnect.

In the case of this evidence reactor, we are already sifting through the more mild cases of errors such that this is almost only triggered when someone has been intentionally mischievous so most likely we should disconnect from them

Reply:
Yeah, something like that -- the consensus reactor similarly "grades" peers as they do (or don't do) useful work. So we should probably have some sort of peer quality scheme that takes all of this into account.

Reply:
I'm actually going to leave this as `PeerErrorSeverityLow` for now until the router is implemented. Then we will revisit all the places we send `PeerError` to ensure the various severities make sense.